scripts/training/finetune-ldm-color-palette.py

```diff
@@ -0,0 +1,322 @@
+import hashlib
+import os
+from dataclasses import dataclass
+from functools import cached_property
+from random import random
+from typing import Any
+
+import requests
+from loguru import logger
+from PIL import Image
+from pydantic import BaseModel
+from torch import Tensor, cat, float32, randn, tensor
+from torch.utils.data import Dataset
+from tqdm import tqdm
+
+import refiners.fluxion.layers as fl
+from refiners.fluxion.adapters.color_palette import ColorPaletteEncoder, SD1ColorPaletteAdapter
+from refiners.fluxion.utils import save_to_safetensors
+from refiners.foundationals.latent_diffusion import (
+    DPMSolver,
+    StableDiffusion_1,
+)
+from refiners.training_utils.callback import Callback
+from refiners.training_utils.huggingface_datasets import HuggingfaceDatasetConfig
+from refiners.training_utils.latent_diffusion import (
+    FinetuneLatentDiffusionConfig,
+    LatentDiffusionConfig,
+    LatentDiffusionTrainer,
+    TestDiffusionConfig,
+    TextEmbeddingLatentsDataset,
+)
+
+
+class ColorPaletteConfig(BaseModel):
+    model_dim: int
+    trigger_phrase: str = ""
+    use_only_trigger_probability: float = 0.0
+    max_colors: int
+
+
+class ColorPalettePromptConfig(BaseModel):
+    text: str
+    color_palette: list[list[float]]
+
+
+class TestColorPaletteConfig(TestDiffusionConfig):
+    prompts: list[ColorPalettePromptConfig] = []
+
+
+class ColorPaletteDatasetConfig(HuggingfaceDatasetConfig):
+    local_folder: str = "data/color-palette"
+
+
+@dataclass
+class TextEmbeddingColorPaletteLatentsBatch:
+    text_embeddings: Tensor
+    latents: Tensor
+    color_palette_embeddings: Tensor
+
+
+class ColorPaletteDataset(TextEmbeddingLatentsDataset):
+    def __init__(
+        self,
+        trainer: "ColorPaletteLatentDiffusionTrainer",
+    ) -> None:
+        super().__init__(trainer=trainer)
+        self.trigger_phrase = trainer.config.color_palette.trigger_phrase
+        self.use_only_trigger_probability = trainer.config.color_palette.use_only_trigger_probability
+        logger.info(f"Trigger phrase: {self.trigger_phrase}")
+        self.color_palette_encoder = trainer.color_palette_encoder
+
+        self.local_folder = trainer.config.dataset.local_folder
+
+        # Download images
+        # Question : there might be a more efficient way to do this
+        # I didn't find the way to do this easily with hugging face
+        # dataset library
+        for item in tqdm(self.dataset, desc="Downloading images"):
+            self.download_image(item)
```

Comment:
What is the best practice to handle HF 'url' dataset ?

Reply:
It's always a bit annoying to use direct URL datasets. The most sensible thing is to download the dataset as an image dataset beforehand. But with some hacks, you can make it work by just downloading the image for each step, and if the download fails, you send an empty batch (you need to modify Trainer's loop a bit to make it work)